cmd/kubeadm/app/util/etcd/etcd.go

```diff
@@ -362,8 +362,11 @@ func (c *Client) RemoveMember(id uint64) ([]Member, error) {
 
 	// Returns the updated list of etcd members
 	ret := []Member{}
-	for _, m := range resp.Members {
-		ret = append(ret, Member{Name: m.Name, PeerURL: m.PeerURLs[0]})
+	if resp != nil {
```

Comment:
the code change here LGTM, but I don't understand why it cannot find the member ID from the endpoint,  the [log](https://storage.googleapis.com/kubernetes-jenkins/logs/ci-kubernetes-e2e-kubeadm-kinder-1-26-on-1-25/1687241264644231168/build-log.txt) doesn't provide any clue for this, and there is no such code to remove this member in any other function from kubeadm? (maybe I miss something)

it is rare to see this error from the serial of test, I guess it might be related to etcd itself, maybe the member is not actual removed but just something wrong happened in the etcd.

Reply:
we wrap the assignment of resp in a backoff. i think there could never be a case where resp is nil, but the err returned from the backoff is not nil.

i guess my point is that resp == nil cannot happen after this check:

```
if err != nil {
	return nil, lastError
}
```

i could be missing something, but to me it seems the change is not needed.

@SataQiu PTAL as well if you have some time.

Reply:
/cc @dlipovetsky 

> Code was added in kubeadm: Make etcd member removal idempotent #117724

Reply:
@dlipovetsky yet
https://github.com/kubernetes/kubeadm/issues/2912 seems like an actual problem

Reply:
```go
if errors.Is(rpctypes.ErrMemberNotFound, err) {
	klog.V(5).Infof("Member was already removed, because member %s was not found", strconv.FormatUint(id, 16))
	return true, nil
}
```
When err is `ErrMemberNotFound`, resp will be nil.

Reply:
ok, it seems return of `true, nil` means we can no longer process resp.
so @pacoxu logic is correct. @chendave can you please check of you have other concerns?

Reply:
the response can be nil only in case that member cannot be found or it's already be removed. 

https://github.com/etcd-io/etcd/blob/main/server/etcdserver/server.go#L165-L168

>	// RemoveMember attempts to remove a member from the cluster. It will
	// return ErrIDRemoved if member ID is removed from the cluster, or return
	// ErrIDNotFound if member ID is not in the cluster.
	RemoveMember(ctx context.Context, id uint64) ([]*membership.Member, error)

take a look at this as well, 
https://github.com/etcd-io/etcd/blob/c1b96473dbdc2b8fa4401d7e00ff57bdbd89b66e/client/v3/cluster.go#L101

> func (c *cluster) MemberRemove(ctx context.Context, id uint64) (*MemberRemoveResponse, error) {
	r := &pb.MemberRemoveRequest{ID: id}
	resp, err := c.remote.MemberRemove(ctx, r, c.callOpts...)
	if err != nil {
		return nil, toErr(ctx, err)
	}
	return (*MemberRemoveResponse)(resp), nil
}



My question how we can hit the situation given that we haven't removed it in somewhere else. 

But for whatever the reason this does happen, then I agree this change is fine.


And the alternative is, we can return the error instead of the `nil` from here as the error is not expected unless it's clear  there is somewhere the member is remove already.

```
		if errors.Is(rpctypes.ErrMemberNotFound, err) {
			klog.V(5).Infof("Member was already removed, because member %s was not found", strconv.FormatUint(id, 16))
			return true, nil     // return true, err instead!!!
		}
````

Reply:
well, looks like this wasn't on hold and it merged. we should backport it to 1.28 and older releases. 

> And the alternative is, we can return the error instead of the nil from here as the error is not expected unless it's clear there is somewhere the member is remove already.

i don't have all the answers but IIRC, this was also the old behavior - i.e. don't return error if member was already removed.